### PR TITLE
fix bugs in setting plane orientation with nD data 

### PR DESCRIPTION
**Pull request by alisterburt**
# Description
fix for #4196 - added some extra guards which ensure that new plane normal directions are always 3D with nD data

cc @brisvag @jni 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Review by brisvag — APPROVED** [review]
As far as I can tell, everything works fine now! I like the additional tests :)

Can you clarify why you're passing `[-3, -2, -1]` rather than `layer._dims_displayed`?

**Comment by alisterburt** [comment]
Thanks! The view direction isn't layer data so it doesn't make sense to slice it like layer data - it still gets promoted to 4d internally in the `layer._world_to_data_ray()` method so needs to be demoted according to the same rules. I remember not liking this 'nD' view direction concept when we introduced it, probably needs to be revisited at some point :)

File: napari/layers/image/_image_key_bindings.py [hunk 1]
```diff
@@ -38,7 +38,9 @@ def orient_plane_normal_along_view_direction(layer: Image):
     viewer = napari.viewer.current_viewer()
     if viewer.dims.ndisplay != 3:
         return
-    layer.plane.normal = layer._world_to_data_ray(viewer.camera.view_direction)
+    layer.plane.normal = layer._world_to_displayed_data_ray(
+        viewer.camera.view_direction, [-3, -2, -1]
```
Comment (jni) [inline napari/layers/image/_image_key_bindings.py]: Can you use a keyword argument here? I have no idea reading this what [-3, -2, -1] is...

**Comment by jni** [comment]
@alisterburt I've left a small code suggestion. I agree with @brisvag though, it's not clear to me why sometimes you use last three axes and sometimes you use layer._dims_displayed. Have you tested this with rolled dimension order?

**Comment by jni** [comment]
(And could you maybe add a comment to those lines to explain what it's doing?)

**Comment by alisterburt** [comment]
added the kwargs and agree it's non obvious - I'll have a go at an explanation but don't know if I can explain in a useful way in code comments 🙁

the `Layer.world_to_data_ray()` method calls `Layer.world_to_data()`

https://github.com/napari/napari/blob/c2058599bee01cdbe0bc247ec4f72ef93929b2a5/napari/layers/base/base.py#L1255-L1269

`Layer.world_to_data()` promotes to the ndim of the layer automagically by prepending zeros

https://github.com/napari/napari/blob/c2058599bee01cdbe0bc247ec4f72ef93929b2a5/napari/layers/base/base.py#L1192-L1195

so our 3D vector (which is a vector in world space, has not been sliced/transformed by the layer/dims) has a zero prepended to it at some point - taking the last three dims just ensures that this operation is 'undone' correctly. Other things such as the cursor position are sliced/transformed by the dims/layer dims so this needs to use the `Layer._dims_displayed()`

the interaction of layer slicing with dims slicing makes this stuff super tricky, I'm hesitant to poke the bear during rc (even if I helped to create it)

I'm super low on bonus time the next few days so can't guarantee I'll get back to this before Sunday, sorry! 😞

**Comment by jni** [comment]
Hahaha calling it a bear is :ok_hand: 

It's definitely iffy but it's not your fault :joy: and I understand the workaround better now. Having this rationale written down here is good. :pray:

**Pull request merged**